src/util/metadata/side_metadata/global.rs

```diff
@@ -97,6 +99,515 @@ impl SideMetadataSpec {
+    ///
+    /// # Safety
+    ///
+    /// This is unsafe because:
+    ///
+    /// 1. Concurrent access to this operation is undefined behaviour.
+    /// 2. Interleaving Non-atomic and atomic operations is undefined behaviour.
+    #[inline(always)]
+    pub unsafe fn store<T: MetadataValue>(&self, data_addr: Address, metadata: T) {
+        self.side_metadata_access::<T, _, _, _>(
+            data_addr,
+            || {
+                let meta_addr = address_to_meta_address(self, data_addr);
+                let bits_num_log = self.log_num_of_bits;
+                if bits_num_log < 3 {
+                    let lshift = meta_byte_lshift(self, data_addr);
+                    let mask = meta_byte_mask(self) << lshift;
+                    let old_val = meta_addr.load::<u8>();
+                    let new_val = (old_val & !mask) | (metadata.to_u8().unwrap() << lshift);
+
+                    meta_addr.store::<u8>(new_val);
+                } else {
+                    meta_addr.store::<T>(metadata);
+                }
+            },
+            |_| {
+                #[cfg(feature = "extreme_assertions")]
+                sanity::verify_store(self, data_addr, metadata);
+            },
+        )
+    }
+
+    #[inline(always)]
+    pub fn load_atomic<T: MetadataValue>(&self, data_addr: Address, order: Ordering) -> T {
+        self.side_metadata_access::<T, _, _, _>(
+            data_addr,
+            || {
+                let meta_addr = address_to_meta_address(self, data_addr);
+                let bits_num_log = self.log_num_of_bits;
+                if bits_num_log < 3 {
+                    let lshift = meta_byte_lshift(self, data_addr);
+                    let mask = meta_byte_mask(self) << lshift;
+                    let byte_val = unsafe { meta_addr.atomic_load::<AtomicU8>(order) };
+                    FromPrimitive::from_u8((byte_val & mask) >> lshift).unwrap()
+                } else {
+                    unsafe { T::load_atomic(meta_addr, order) }
+                }
+            },
+            |_v| {
+                #[cfg(feature = "extreme_assertions")]
+                sanity::verify_load(self, data_addr, _v);
+            },
+        )
+    }
+
+    #[inline(always)]
+    pub fn store_atomic<T: MetadataValue>(&self, data_addr: Address, metadata: T, order: Ordering) {
+        self.side_metadata_access::<T, _, _, _>(
+            data_addr,
+            || {
+                let meta_addr = address_to_meta_address(self, data_addr);
+                let bits_num_log = self.log_num_of_bits;
+                if bits_num_log < 3 {
+                    let lshift = meta_byte_lshift(self, data_addr);
+                    let mask = meta_byte_mask(self) << lshift;
+                    let metadata_u8 = metadata.to_u8().unwrap();
+                    let mut old_val = unsafe { meta_addr.load::<u8>() };
+                    let mut new_val = (old_val & !mask) | (metadata_u8 << lshift);
```

Comment:
Same here.  I recommend `debug_assert!` that `metadata_u8` does not have more than `bits_num_log` significant bits.